### DISCUSSION TITLE
What is WebGPU and how does that effect WebGL?

### DISCUSSION THREAD

**Discussion by lv2lrn4lf** [gfxfundamentals/webgl-fundamentals#398] [General]
I been hearing and seeing people and content saying webgl is dead and webGPU is the new wegbl and will replace webgl down the line... is this true? Honestly i kinda know that they removing webgl support is as likely as they start charging people to user web browsers... very unlikely seeing that the web value backwards compatibility, but i wanted to know what the people who manage webgl sites.

**Comment by greggman** [comment]
There are no plans to remove WebGL just like there are no plans to remove XMLHTTPRequest, even though XMLHTTPRequest is deprecated. Too many sites would break if they removed WebGL.

WebGPU has not shipped yet. The spec is not finished. The API and the shading language are still changing. WebGPU requires higher end GPUs than WebGL so less users will be able to use it. It's also arguably harder to use in many ways.

But, yes, WebGL2 will be the last WebGL version and over time most sites that want to do GPU stuff will likely switch to WebGPU over the next 5-6 years.

Note that most of the techniques on this site are the same techniques you'd use in WebGPU. You still make vertex shaders, they still output a position. You still make fragment shaders, they still output a color. You still give them the same data (buffers, attributes, unifiorms, textures). What changes is the API calls you use to set things up and you write shaders in WGSL instead of GLSL

**Reply by lv2lrn4lf** [reply]
Just to make sure wgsl is glsl just in webgl runtime?

**Reply by greggman** [reply]
Each graphics API has it's own shading language (the language you write shaders in)

WGSL is the shading language of WebGPU (**W**eb**G**PU **S**hading **L**anguage)

GLSL is the shading language of OpenGL/OpenGLES/WebGL

HLSL is the shading language of Direct3D

MSL is the shading language of Metal

SPIR-V is the shading language of Vulkan

CG is NVidia's custom shading language

and there are others

**Reply by lv2lrn4lf** [reply]
I heard of the vulkan and metal on twitch and look them up.. so these language varies from each other, they different. that suck.... so i'm not writing gls, i'm writing wgsl webgl shader language. Been cool if we can carry the shader code from one to another.

**Reply by greggman** [reply]
[wgsl](https://www.w3.org/TR/WGSL/) is webgpu shading language, **not** WebGL shading language. WebGL uses GLSL. In particular, WebGL uses [GLSL ES 1.0](https://www.khronos.org/files/opengles_shading_language.pdf). WebGL2 can use both GLSL ES 1.0 and [GLSL ES 3.0](https://www.khronos.org/registry/OpenGL/specs/es/3.0/GLSL_ES_Specification_3.00.pdf)

**Comment by greggman** [comment]
If you'd like to see some differences between WebGL and WebGPU there's this article

https://webgpufundamentals.org/webgpu/lessons/webgpu-from-webgl.html

The site has no other articles yet and, like I mentioned above, the spec is still changing. Once it's settled down I'll look into converting the applicable WebGL articles to WebGPU

**Reply by lv2lrn4lf** [reply]
I skim through it, look at some parts more closely than others... I see that with webgpu you don't retrieve it from the canvas object, and that webgl use glsl while webgpu use wgsl which seem to be a little more robust. I kinda wish that they kept them both glsl instead of making it different.. i don't see a issue with them making the api exactly the same or maybe the same but with hidding a lot of steps... or just making it exactly the same api as webgl period.... that would had been nice in my opinion.
